File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2021/Y21D24.kt
package org.gristle.pdxpuzzles.advent.y2021

import org.gristle.pdxpuzzles.advent.utilities.Day
import java.util.*
import kotlin.math.max
import kotlin.math.min

class Y21D24(input: String) : Day {
    /**
     * Input boils down to an 18 line function run 14 times. The function has two parameters, x and y, and
     * the algorithm depends on whether x is positive. 7 of these have a positive x value and 7 have a
     * negative x value. When x is positive, z := 26 * z + y + the supplied digit. Thus, z increases. When x
     * is negative, if (z % 26) + x equals the supplied digit, z := z / 26, rounded down to nearest integer. If
     * it doesn't equal the supplied digit, z becomes an even larger number.
     *
     * In order for the final value to be 0, every step with a negative x must result in division by 26. Thus,
     * when such a step is encountered, the z value must be such that (z % 26) + x is a number between 1..9. We
     * use a stack to keep track of what z will have to be for the negative x steps. We pair the z-increasing
     * functions with the z-decreasing functions, and find shared values that will satisfy both.
     */

    // Tuple to carry step information.
    data class Step(val order: Int, val x: Int, val y: Int)

    // Tuple to hold "push" and "pop" steps together. 
    data class PairedSteps(val push: Step, val pop: Step)

    // Take the input, glean the parameters to the function, and pair the "push" and "pop" steps using a stack. 
    private val pairedSteps = input
        .lines()
        .chunked(18) // Split the program into the 18-line repeating function
        .mapIndexed { index, lines ->
            val xInc = lines[5].substringAfterLast(" ").toInt() // Extract x
            val yInc = lines[15].substringAfterLast(" ").toInt() // Extract y
            Step(index, xInc, yInc) // Map to a step
        }.let { steps -> // Take those steps and pair them together.
            buildList<PairedSteps> {
                val orderStack: Deque<Step> = ArrayDeque() // We will use a stack to pair the steps.
                steps.forEach { step ->
                    if (step.x >= 0) { // If x is positive, this is a "push" step.
                        orderStack.push(step)
                    } else { // If x is negative, pop the last "pushed" step, and pair the two together.
                        add(PairedSteps(orderStack.pop(), step))
                    }
                }
                sortBy { it.push.order } // Sort by the order of the "push" steps so that z is increased properly
            }
        }

    fun solve(findIntersection: (pushMax: Int, popMax: Int) -> Int): Long {
        var z = 0L

        // The steps are paired and are run in the order that the "push" steps occur. This presents a problem
        // because the "pull" steps can occur out of order. However, we know that the "pull" steps always result
        // in z:= z / 26. So this kludge allows us to figure how many intervening "pull" steps occurred since the
        // last "push" step, and reduce z accordingly.
        var lastIncrease = 0

        // This is our answer, an array of digits. It is a mutable list because the digits are not found in order. 
        val modelNumber = MutableList(pairedSteps.size * 2) { 0 }

        pairedSteps.forEach { step ->
            val increaseZ = step.push.y + 26 * z // The amount z will increase by, apart from the supplied digit
            val pushMax = ((increaseZ + 9) % 26).toInt() // The highest amount z can increase by, mod 26.
            val popMax = -step.pop.x + 9 // The highest allowed value for (z % 26) that still results in z := z / 26

            /* 
             * If we are looking for the highest number, we grab the minimum between the pushMax and the popMax.
             * Those numbers represent the two individual maximums; by taking the minimum between the two, we grab
             * the largest number that works for both.
             * 
             * If we are looking for the lowest number, we grab the maximum between the pushMax and the popMax, 
             * then subtract by 8. The subtraction effectively makes it a comparison between the pushMin and the
             * popMin. The maximum between these two represents the smallest number that works for both.
             */
            val intersection = findIntersection(pushMax, popMax)

            // Finds the digit corresponding to the intersection z value for both push and pop places. 
            modelNumber[step.push.order] = 9 - (pushMax - intersection)
            modelNumber[step.pop.order] = 9 - (popMax - intersection)

            // Applies the z-decrease step for the number of times that a pop had occured since the last push
            repeat(step.push.order - lastIncrease - 1) { z /= 26 }

            // Applies the z-increase step.
            z = increaseZ + modelNumber[step.push.order]
            lastIncrease = step.push.order
        }

        return modelNumber.joinToString("").toLong()
    }

    override fun part1() = solve { pushMax, popMax -> min(pushMax, popMax) }

    override fun part2() = solve { pushMax, popMax -> max(pushMax, popMax) - 8 }
}

fun main() = Day.runDay(Y21D24::class)
//    var time = System.nanoTime()
//    val c = Y21D24(readRawInput("y2021/d24"))
//    println("Class creation: ${elapsedTime(time)}ms")
//    time = System.nanoTime()
//    println("Part 1: ${c.part1()} (${elapsedTime(time)}ms)") // 92969593497992
//    time = System.nanoTime()
//    println("Part 2: ${c.part2()} (${elapsedTime(time)}ms)") // 81514171161381
//}

File: rust/advent/src/bin/2021/y21d24.rs
/**
 * Input boils down to an 18 line function run 14 times. The function has two parameters, x and y, and
 * the algorithm depends on whether x is positive. 7 of these have a positive x value and 7 have a
 * negative x value. When x is positive, z := 26 * z + y + the supplied digit. Thus, z increases. When x
 * is negative, if (z % 26) + x equals the supplied digit, z := z / 26, rounded down to nearest integer. If
 * it doesn't equal the supplied digit, z becomes an even larger number.
 *
 * In order for the final value to be 0, every step with a negative x must result in division by 26. Thus,
 * when such a step is encountered, the z value must be such that (z % 26) + x is a number between 1..9. We
 * use a stack to keep track of what z will have to be for the negative x steps. We pair the z-increasing
 * functions with the z-decreasing functions, and find shared values that will satisfy both.
 */

use std::cmp::{max, min};

use advent::utilities::get_input::get_input;
use itertools::Itertools;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input = Vec<PairedSteps>;
type Output = i64;

struct PairedSteps {
    push: Step,
    pop: Step,
}

struct Step {
    order: usize,
    x: i64,
    y: i64,
}

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(21, 24).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    let steps = input.lines().chunks(18).into_iter().enumerate()
        .map(|(index, lines)| {
            let lines = lines.collect_vec();
            let x_inc = lines[5];
            let x_inc: i64 = x_inc[x_inc.as_bytes().iter().rposition(|&b| b == b' ').unwrap() + 1 ..]
                .parse()
                .unwrap();
            let y_inc = lines[15];
            let y_inc: i64 = y_inc[y_inc.as_bytes().iter().rposition(|&b| b == b' ').unwrap() + 1 ..]
                .parse()
                .unwrap();
            Step { order: index, x: x_inc, y: y_inc }
        })
        .collect_vec();

    let mut paired_steps = Vec::new();
    let mut order_stack: Vec<Step> = Vec::new();

    for step in steps {
        if step.x >= 0 {
            order_stack.push(step);
        } else {
            paired_steps.push(PairedSteps { push: order_stack.pop().unwrap(), pop: step })
        }
    }
    
    paired_steps.sort_by_key(|it| it.push.order);
    paired_steps
}

fn solve<F>(paired_steps: &Input, find_intersection: F) -> Output
where
    F: Fn(i64, i64) -> i64,
{
    let mut z = 0;
    let mut model_number = vec![0; paired_steps.len() * 2];

    for step in paired_steps {
        let increase_z = step.push.y + 26 * z;
        let push_max = (increase_z + 9) % 26 ;
        let pop_max = -step.pop.x + 9;

        let intersection = find_intersection(push_max, pop_max);

        model_number[step.push.order] = 9 - (push_max - intersection);
        model_number[step.pop.order] = 9 - (pop_max - intersection);
        
        z = increase_z + model_number[step.push.order];
    }

    model_number.into_iter().reduce(|acc, n| acc * 10 + n).unwrap()
}

fn part1(paired_steps: &Input) -> Output {
    solve(paired_steps, |push_max, pop_max| min(push_max, pop_max))
}

fn part2(paired_steps: &Input) -> Output {
    solve(paired_steps, |push_max, pop_max| max(push_max, pop_max) - 8)
}

#[test]
fn default() {
    let input = get_input(21, 24).unwrap();
    let input = parse_input(&input);
    assert_eq!(92969593497992, part1(&input));
    assert_eq!(81514171161381, part2(&input));
}

// Input parsed (29μs)
// 1. 92969593497992 (9μs)
// 2. 81514171161381 (2μs)
// Total: 44μs